### PR TITLE
Use crypt_gensalt(), if available in libcrypt.

### PR DESCRIPTION
**Pull request by besser82**
Most Linux distributions, including Fedora and RHEL 8, are shipping with libxcrypt >= 4.0.

Since that version of libxcrypt the provided family of crypt_gensalt() functions are able to use automatic entropy drawn from secure system ressources, like arc4random(), getentropy() or getrandom().

Anyways, the settings generated by crypt_gensalt() are always guaranteed to works with the crypt() function.

Using crypt_gesalt() is also needed to make proper use of newer hashing methods, like yescrypt, provided by libxcrypt.

***

See https://github.com/shadow-maint/shadow/pull/303#issuecomment-764847950 for reference.

### CODE REVIEW

**Comment by besser82** [comment]
@solardiz: This is PR going to make you happy.  =)

File: libmisc/salt.c [hunk 1]
```diff
@@ -12,19 +12,314 @@
...
+		}
+		setting[SETTING_SIZE - 1] = '\0';
+		fprintf (shadow_logfd,
+			 _("Invalid ENCRYPT_METHOD value: '%s'.\n"
+			   "Defaulting to DES.\n"),
+			 method);
+	}
+
+	crypt_gensalt_rn(setting, rounds, NULL, 0,
+			 result, CRYPT_GENSALT_OUTPUT_SIZE);
```
Comment (solardiz) [inline libmisc/salt.c:310]: Let's check the return value here. Return value is documented, what's in `result` upon error is not, and we're not in the same library with the implementation to make such assumptions.

**Review by solardiz — COMMENTED** [review]
While I like where this is going (thank you!), I think it's just a first step, and I find it weird that with this PR's changes we at the same time require `crypt_gensalt_rn`, yet still have the old salt generation code somewhere else in shadow (dead code now?), the legacy per-hash-type settings, and the random rounds count in a range nonsense.

**Comment by solardiz** [comment]
As you can see from the Owl and ALT patches I referenced in #303, longer-term there should be only two settings - `CRYPT_PREFIX` and `CRYPT_ROUNDS` - and nothing per-hash-type.

I understand this might not be right to "force" onto shadow users without warning (having their old settings stop working). However, in that case for consistency perhaps the dependency on libxcrypt 4.0+ should be optional for now?

Just thinking out loud.

**Comment by solardiz** [comment]
Let's also hear from @ldv-alt on this.

File: libmisc/salt.c (hunk 1 above)
Comment (besser82) [reply libmisc/salt.c:310]: Fixed in rebased commit.

**Comment by besser82** [comment]
@solardiz: The changes, I applied in this PR, are meant to be as transparent as possible to the end-user and/or distribution maintainer:

  * `crypt_gensalt()` is only used, if the system provides libxcrypt >= 4.0, otherwise the old code-path will be used.
  * There are no changes needed for existing configurations and such.

File: libmisc/salt.c [hunk 2]
```diff
@@ -413,7 +645,7 @@ static /*@observer@*/const char *gensalt (size_t salt_size)
 #ifdef USE_YESCRYPT
 	} else if (0 == strcmp (method, "YESCRYPT")) {
 		MAGNUM(result, 'y');
-		strcat(result, YESCRYPT_salt_cost((int *)arg));
+		strcat(result, YESCRYPT_salt_cost((long *)arg));
```
Comment (solardiz) [inline libmisc/salt.c:648]: Looks like this was actually a bug being fixed here. So at least this bug fix should get in even if this PR maybe doesn't.

File: libmisc/salt.c [hunk 3]
```diff
@@ -12,19 +12,321 @@
...
+	}
+
+	retval = crypt_gensalt_rn(setting, rounds, NULL, 0,
+				  result, CRYPT_GENSALT_OUTPUT_SIZE);
+
+	if (retval == NULL) {
+		for (int i = 0; i < CRYPT_GENSALT_OUTPUT_SIZE; ++i) {
+			result[i] = '\0';
+		}
+	}
```
Comment (solardiz) [inline libmisc/salt.c:321]: Why not call `crypt_gensalt_rn` from inside the `if`, so that we wouldn't need to introduce a variable? And why not use `memset` (if zeroizing the entire buffer is somehow needed, which I doubt)?

File: libmisc/salt.c [hunk 4]
```diff
@@ -12,19 +12,321 @@
...
+
+	return result;
+}
+
+#else
```
Comment (solardiz) [inline libmisc/salt.c:326]: Maybe add ` /*  USE_XCRYPT_GENSALT */` here, although I don't know if this project's style requires that we possibly negate the condition in such comment on `#else`.

File: libmisc/salt.c (hunk 2 above)
Comment (besser82) [reply libmisc/salt.c:648]: Indeed!

File: libmisc/salt.c (hunk 3 above)
Comment (besser82) [reply libmisc/salt.c:321]: Fixed in rebased commit.

File: libmisc/salt.c (hunk 4 above)
Comment (besser82) [reply libmisc/salt.c:326]: Fixed in rebased commit.

File: libmisc/salt.c (hunk 2 above)
Comment (besser82) [reply libmisc/salt.c:642]: I've opened https://github.com/shadow-maint/shadow/pull/358 to keep this in mind, if the current PR doesn't get merged.

**Comment by solardiz** [comment]
BTW, why not use the non-thread-safe `crypt_gensalt` if we put the result in a static buffer anyway?

**Comment by besser82** [comment]
> BTW, why not use the non-thread-safe `crypt_gensalt` if we put the result in a static buffer anyway?

Wouldn't that just reorder zeroing the static buffer, need another variable for error checking, and require to strncpy the result of `crypt_gensalt` into the static buffer on success?

**Comment by solardiz** [comment]
I was thinking that we'd simply return the pointer returned by `crypt_gensalt`, with no extra buffer nor copying. As to zeroing the buffer on failure, why? If we do that because we're concerned about any sensitive information being left in there (what information? just the hash type/settings?) then it's better done inside libxcrypt, not in the callers.

**Comment by besser82** [comment]
> I was thinking that we'd simply return the pointer returned by `crypt_gensalt`, with no extra buffer nor copying.

Okay, that makes sense.  Changed that in rebased commit.

**Review by ikerexxe — CHANGES_REQUESTED** [review]
I've done a review of the c code and noted several items. Can you improve them?

By the way, I've seen that https://github.com/shadow-maint/shadow/pull/358 and this PR have several changes in common. Can you consider removing the common changes from this PR?

**Comment by solardiz** [comment]
@ikerexxe I think most of your comments are about code/issues copied from elsewhere in the original file. My thinking was that improving this old code is out of scope of this PR, and increasing differences between the original vs. copied instances isn't obviously an improvement. Ideally, all of this would be obsoleted with more complete use of `crypt_gensalt` later, as I had suggested. At least the random rounds number in a range nonsense should be dropped soon, I think with a separate PR.

A possible improvement to this PR would be to avoid such code copying/duplication, if this can be done cleanly.

**Comment by ldv-alt** [comment]
> Let's also hear from @ldv-alt on this.

ALT is using Owl's crypt_gensalt-based approach for almost 20 years now.
In ALT's shadow package [libmisc/salt.c](http://git.altlinux.org/gears/s/..git?p=shadow.git;a=blob;f=shadow/libmisc/salt.c) contains essentially the following (everything else is #if 0'ed):
```
const char *
crypt_make_salt(/*@null@*/const char *meth, /*@null@*/void *arg)
{
        char *retval;

        retval = crypt_gensalt(getdef_str("CRYPT_PREFIX"),
                        getdef_num("CRYPT_ROUNDS", 0), NULL, 0);
        if (!retval) {
                fprintf(stderr, "Unable to generate a salt, "
                                "check your CRYPT_PREFIX and CRYPT_ROUNDS settings.\n");
                exit(1);
        }

        return retval;
}
```

So from my point of view, the sooner all this legacy code is gone the better.

**Comment by besser82** [comment]
> I've done a review of the c code and noted several items. Can you improve them?

Working on it.


> By the way, I've seen that #358 and this PR have several changes in common. Can you consider removing the common changes from this PR?

This is intentional, as I've split out a bugfix into #358, that should not be overlooked in case this PR will *not* get merged.

**Comment by solardiz** [comment]
I think if you keep the bugfix in this PR as well, then it should be a separate first commit (same as what's in the other PR).

File: libmisc/salt.c [hunk 5]
```diff
@@ -12,19 +12,309 @@
...
+/*
+ * Generate 8 base64 ASCII characters of random salt.  If MD5_CRYPT_ENAB
+ * in /etc/login.defs is "yes", the salt string will be prefixed by "$1$"
+ * (magic) and pw_encrypt() will execute the MD5-based FreeBSD-compatible
+ * version of crypt() instead of the standard one.
+ * Other methods can be set with ENCRYPT_METHOD
+ *
+ * The method can be forced with the meth parameter.
+ * If NULL, the method will be defined according to the MD5_CRYPT_ENAB and
+ * ENCRYPT_METHOD login.defs variables.
```
[the remark this answers is not in the document]
Comment (besser82) [reply libmisc/salt.c:260]: Clarified in rebased commit.

File: libmisc/salt.c [hunk 6]
```diff
@@ -12,19 +12,309 @@
...
+ * If meth is specified, an additional parameter can be provided.
+ *  * For the SHA256 and SHA512 method, this specifies the number of rounds
+ *    (if not NULL).
+ *  * For the YESCRYPT method, this specifies the cost factor (if not NULL).
+ */
+/*@observer@*/const char *crypt_make_salt (/*@null@*//*@observer@*/const char *meth, /*@null@*/void *arg)
+{
+	const char *method;
+	char setting[SETTING_SIZE];
+        unsigned long rounds = 0;
```
[the remark this answers is not in the document]
Comment (besser82) [reply libmisc/salt.c:271]: Fixed in rebased commit.

File: libmisc/salt.c [hunk 7]
```diff
@@ -12,19 +12,309 @@
...
+	if (cost > Y_COST_MAX) {
+		cost = Y_COST_MAX;
+	}
+
+	return cost;
+}
+#endif /* USE_YESCRYPT */
+
+/* Maximum size of the setting for crypt_gensalt(). */
+#define SETTING_SIZE 16
```
[the remark this answers is not in the document]
Comment (besser82) [reply libmisc/salt.c:249]: Fixed in rebased commit.

File: libmisc/salt.c [hunk 8]
```diff
@@ -12,19 +12,309 @@
...
+		}
+	} else {
+		rounds = (unsigned long) *prefered_rounds;
+	}
+
+	/*
+	 * Sanity checks.
+	 * Use 19 as an upper bound for now,
+	 * because musl doesn't allow rounds >= 20.
+	 */
```
[the remark this answers is not in the document]
Comment (besser82) [reply libmisc/salt.c:209]: Removed, as obviously wrong, in rebased commit.

File: libmisc/salt.c [hunk 9]
```diff
@@ -12,19 +12,309 @@
...
+			max_rounds = min_rounds;
+		}
+
+		if (min_rounds > max_rounds) {
+			max_rounds = min_rounds;
+		}
+
+		rounds = (unsigned long) shadow_random (min_rounds, max_rounds);
+	} else if (0 == *prefered_rounds) {
+		return 0;
```
[the remark this answers is not in the document]
Comment (besser82) [reply libmisc/salt.c:152]: Removed, as unneeded, in rebased commit.

File: libmisc/salt.c [hunk 10]
```diff
@@ -12,19 +12,309 @@
...
+	 * Sanity checks.
+	 * Use 19 as an upper bound for now,
+	 * because musl doesn't allow rounds >= 20.
+	 */
+	if (rounds < B_ROUNDS_MIN) {
+		rounds = B_ROUNDS_MIN;
+	}
+
+	if (rounds > B_ROUNDS_MAX) {
+		/* rounds = B_ROUNDS_MAX; */
```
[the remark this answers is not in the document]
Comment (besser82) [reply libmisc/salt.c:215]: This is still needed, as `B_ROUNDS_MAX` is defined to `31`, which is really the maximum rounds of the bcrypt hash method.

**Comment by besser82** [comment]
> I think if you keep the bugfix in this PR as well, then it should be a separate first commit (same as what's in the other PR).

Will fix it that way.

File: libmisc/salt.c [hunk 11]
```diff
@@ -12,19 +12,302 @@
...
+
+/*
+ * Add the salt prefix.
+ */
+#define MAGNUM(array,ch)	(array)[0]=(array)[2]='$',(array)[1]=(ch),(array)[3]='\0'
+#ifdef USE_BCRYPT
+/*
+ * Using the Prefix $2b$ to enable an anti-collision safety measure in musl libc.
+ * Negatively affects a subset of passwords containing the '\xff' character,
+ * which is not valid UTF-8 (so "unlikely to cause much annoyance").
```
Comment (solardiz) [inline libmisc/salt.c:41]: When revised to say `$2b$`, this comment became factually wrong. The "anti-collision safety measure" was only enabled for `$2a$`.

**Comment by solardiz** [comment]
There's a typo in the function name in "Using crypt_gesalt() is also needed" in one of the commit messages.

File: libmisc/salt.c (hunk 11 above)
Comment (solardiz) [reply libmisc/salt.c:41]: To be clear, I think this comment should simply be dropped. In fact, it never made sense to deliberately enable the anti-collision safety measure for new password, like we do in the shadow package. That measure was only intended for authentication against old potentially-buggy hashes, not for new password hashes.

Comment (besser82) [reply libmisc/salt.c:41]: Fixed in rebased commit.

**Comment by besser82** [comment]
> There's a typo in the function name in "Using crypt_gesalt() is also needed" in one of the commit messages.

Also fixed.

File: libmisc/salt.c [hunk 12]
```diff
@@ -90,12 +90,11 @@ static void seedRNG (void)
  */
 #define MAGNUM(array,ch)	(array)[0]=(array)[2]='$',(array)[1]=(ch),(array)[3]='\0'
 #ifdef USE_BCRYPT
-/* 
- * Using the Prefix $2a$ to enable an anti-collision safety measure in musl libc.
- * Negatively affects a subset of passwords containing the '\xff' character,
- * which is not valid UTF-8 (so "unlikely to cause much annoyance").
+/*
+ * Using the Prefix $2b$, which is the preferred prefix to use for new bcrypt
+ * hahses for a long time now.
```
Comment (solardiz) [inline libmisc/salt.c]: Typo in "hahses". I'd just drop the comment, really.

**Comment by solardiz** [comment]
I find it weird that this PR adds as many as ~250 lines of code, most of those being copies of legacy code. There ought to be a clean way to rework it to keep and temporarily reuse the legacy code without having it duplicated. Perhaps the `#if`/`#else`/`#endif` could be moved further down, to be after the legacy functions that we temporarily reuse?

File: libmisc/salt.c (hunk 12 above)
Comment (besser82) [reply libmisc/salt.c]: Fixed in rebased commit.

**Comment by besser82** [comment]
> I find it weird that this PR adds as many as ~250 lines of code, most of those being copies of legacy code. There ought to be a clean way to rework it to keep and temporarily reuse the legacy code without having it duplicated. Perhaps the `#if`/`#else`/`#endif` could be moved further down, to be after the legacy functions that we temporarily reuse?

Done so in rebased commit.

**Comment by besser82** [comment]
@solardiz I've given this a complete overhaul, and split out changes for code cleanup into a seperate commit before doing the implemetation of `crypt_gensalt()`.

Please have another look, if you can spare the time.

**Comment by besser82** [comment]
A breakdown of this PR can be found in the following PRs:

* Small bugfix for yescrypt impl.: https://github.com/shadow-maint/shadow/pull/358
* bcypt should use `$2b$`: https://github.com/shadow-maint/shadow/pull/360
* Cleanup of `libmisc/salt.c`: https://github.com/shadow-maint/shadow/pull/361

**Comment by besser82** [comment]
@ikerexxe Would you mind to have a second look on this?

File: libmisc/salt.c [hunk 13]
```diff
@@ -12,33 +12,98 @@
...
 #include <stdlib.h>
 #include <stdio.h>
 #include <assert.h>
 #include "prototypes.h"
 #include "defines.h"
 #include "getdef.h"
 
+#if (defined CRYPT_GENSALT_IMPLEMENTS_AUTO_ENTROPY && \
+     CRYPT_GENSALT_IMPLEMENTS_AUTO_ENTROPY)
+/* Maximum size of the setting for crypt_gensalt(). */
```
Comment (solardiz) [inline libmisc/salt.c:26]: This comment "Maximum size of the setting for crypt_gensalt()." looks irrelevant here.

**Review by solardiz — COMMENTED** [review]
Great that the line count doesn't increase much anymore (except in the pot file), but now the "Sanitize code" commit makes somewhat invasive changes to legacy code, which I didn't fully review. Hopefully, they're OK.

File: libmisc/salt.c (hunk 13 above)
Comment (besser82) [reply libmisc/salt.c:26]: Ahh, yes, that somehow slipped into there…  Will remove with next push.

Comment (besser82) [reply libmisc/salt.c]: Fixed.

**Comment by ikerexxe** [comment]
> @ikerexxe Would you mind to have a second look on this?

Sure! I've been quite busy these days but I expect to have some time tomorrow.

**Review by ikerexxe — COMMENTED** [review]
The code now is in a better shape and it's easier to understand. I've added several minor concerns but they should be easy to fix.

By the way, taking into account that you are already touching `salt.c` file, do you mind making the necessary changes to fix https://github.com/shadow-maint/shadow/issues/376?

File: libmisc/salt.c [hunk 14]
```diff
@@ -156,196 +203,174 @@ static /*@observer@*/const char *SHA_salt_rounds (/*@null@*/int *prefered_rounds
...
+{
+	const size_t buf_begin = strlen (buf);
+
+	/* Nothing to do here if SHA_ROUNDS_DEFAULT is used. */
+	if (rounds == SHA_ROUNDS_DEFAULT) {
+		return;
+	}
+
+	/* Check if the result buffer is long enough. */
+	assert (GENSALT_SETTING_SIZE > buf_begin + 17);
```
[the remark this answers is not in the document]
Comment (besser82) [reply libmisc/salt.c:250]: Fixed. See #377 also.

File: libmisc/salt.c [hunk 15]
```diff
@@ -156,196 +203,174 @@ static /*@observer@*/const char *SHA_salt_rounds (/*@null@*/int *prefered_rounds
...
+
+	(void) snprintf (buf + buf_begin, 18, "rounds=%lu$", rounds);
```
[the remark this answers is not in the document]
Comment (besser82) [reply libmisc/salt.c:252]: Fixed. See #377 also.

File: libmisc/salt.c [hunk 16]
```diff
@@ -156,196 +203,174 @@ static /*@observer@*/const char *SHA_salt_rounds (/*@null@*/int *prefered_rounds
...
  */
-static /*@observer@*/const char *gensalt_bcrypt (void)
+static /*@observer@*/void BCRYPT_salt_rounds_to_buf (char *buf, unsigned long rounds)
 {
-	static char salt[32];
+	const size_t buf_begin = strlen (buf);
 
-	salt[0] = '\0';
+	/* Check if the result buffer is long enough. */
+	assert (GENSALT_SETTING_SIZE > buf_begin + 3);
```
[the remark this answers is not in the document]
Comment (besser82) [reply libmisc/salt.c:328]: Fixed. See #377 also.

File: libmisc/salt.c [hunk 17]
```diff
@@ -156,196 +203,174 @@ static /*@observer@*/const char *SHA_salt_rounds (/*@null@*/int *prefered_rounds
...
-	seedRNG ();
-	strcat (salt, l64a (random()));
-	do {
-		strcat (salt, l64a (random()));
-	} while (strlen (salt) < BCRYPT_SALT_SIZE);
-
-	salt[BCRYPT_SALT_SIZE] = '\0';
-
-	return salt;
+	(void) snprintf (buf + buf_begin, 4, "%2.2lu$", rounds);
```
[the remark this answers is not in the document]
Comment (besser82) [reply libmisc/salt.c:330]: Fixed. See #377 also.

**Comment by besser82** [comment]
> The code now is in a better shape and it's easier to understand. I've added several minor concerns but they should be easy to fix.

Done!


> By the way, taking into account that you are already touching `salt.c` file, do you mind making the necessary changes to fix #376?

See #377 also.

**Comment by besser82** [comment]
@hallyn Last one and we're done, I promise!  :P

**Comment by ikerexxe** [comment]
LGTM. Thank you for your patches!

**Comment by besser82** [comment]
> LGTM. Thank you for your patches!

Please wait with merging!  I forgot to include <crypt.h>…

**Comment by besser82** [comment]
> > LGTM. Thank you for your patches!
> 
> Please wait with merging! I forgot to include <crypt.h>…

@hallyn Things should be fine for merging now.

**Comment by besser82** [comment]
@solardiz @ikerexxe Could you please set the review flag to approved, so @hallyn knows this PR is good to merge now?

**Comment by solardiz** [comment]
@besser82 I'm not formally "approving" this for two reasons: (1) I am not a maintainer and (2) my review of changes to the legacy code wasn't sufficiently thorough and I don't feel like spending enough of my time on the legacy code that I don't use to perform its proper review and testing (and this really needs retesting of all settings, not just review).

**Review by ikerexxe — COMMENTED** [review]
I have one last concern. Why have you included the changes on the `.po` files? Have you changed anything on those files?

**Comment by besser82** [comment]
> I have one last concern. Why have you included the changes on the `.po` files? Have you changed anything on those files?

No, but I updated the pot file, since my changes have added new translatable strings.  And as the pot file was updated a simple run of `make` now updates the po files as well.  Thus I've included them here, so other developers don't get annoyed by having them listed as modified files in their clone.

**Comment by besser82** [comment]
> > I have one last concern. Why have you included the changes on the `.po` files? Have you changed anything on those files?
> 
> No, but I updated the pot file, since my changes have added new translatable strings. And as the pot file was updated a simple run of `make` now updates the po files as well. Thus I've included them here, so other developers don't get annoyed by having them listed as modified files in their clone.

Anyways, I've dropped the update of the translation files from this PR, and will open a new PR for them.

**Comment by ikerexxe** [comment]
> > > I have one last concern. Why have you included the changes on the `.po` files? Have you changed anything on those files?
> > 
> > 
> > No, but I updated the pot file, since my changes have added new translatable strings. And as the pot file was updated a simple run of `make` now updates the po files as well. Thus I've included them here, so other developers don't get annoyed by having them listed as modified files in their clone.
> 
> Anyways, I've dropped the update of the translation files from this PR, and will open a new PR for them.

Thanks. LGTM!

**Review by ikerexxe — APPROVED** [review]

**Comment by besser82** [comment]
@hallyn This PR should be ready to merge, then.
